---
title: "SLE712 assessment part 1 and part 2"
author: "Kinhgsley Isaac Ogamanya"
date: "18/05/2020"
output: html_document
theme: cosmo
---

# PART 1: Analyse the tsv and csv files

##Question1: Read in the file, making the gene accession numbers the row names. Show a table of values for the first six genes.
```{r,download}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/gene_expression.tsv",
              destfile = "gene_expression.tsv")
```
I use the download.file command to import files from GitHub 
And then i read in R making sure the first columns is set as the row names and headers are recongnized 
so as to show a table of values for the first six genes, and then proceed into using the head command to check if the gene has been imported properly, by showing a table of values for the first six genes and also checking the data structure.
```{r,import_read}
gene <- read.table("gene_expression.tsv",header = TRUE , stringsAsFactors = FALSE , row.names = 1)
head(gene)
str(gene)
```

##Question2: make a new column which is the mean of the other columns. Show a table of values for the first six genes
```{r,chunk1}
gene$mean <- rowMeans(gene)
head(gene)
str(gene)
```
In this question2, i use the rowMeans command to create the mean of the new columns 
and then using the head command to check if the gene has been imported properly 
by showing a table of values for the first six genes and also checking the data structure


##Question3: List the 10 genes with the highest mean expression
```{r,chunk2}
head(gene[order(-gene$mean), ], 10)
```
In question 3, i use the order command and square brackets to reorder the gene files 
and providing their mean and then using the head command to check 
and list genes with the highest mean expression

##Question4: Determine the number of genes with a mean <10
```{r,chunk3}
filtered <- subset(gene, mean < 10)
nrow(filtered)
```
in question 4, i use the subset command to determine the genes with a mean <10
and then filtered to get the total number of genes with a mean <10

##Question5: Make a histogram plot of the mean values in png format and paste it into your report
```{r,plot1}
hist(gene$mean)
```
I created a histogram showing the frequency against the gene$mean at the x-axis

##Question6: Import this csv file into an R object. What are the column names?
```{r,chunk4}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/growth_data.csv",
              destfile = "growth_data.csv")
```
Using the download.file command to import files from GitHub 
And then i read in R using the `read.csv()`to obtain the dataset and also making sure the header are recongnized and seting the `stringsAsFcators = FALSE`and then proceed into using the head command to check if the gene has been imported properly, by showing a table of values for the first six genes and also checking the data structure, then finally use the `colnames` command to list the column names
```{r,chunk5}
tree <- read.csv("growth_data.csv", header = TRUE, stringsAsFactors = FALSE)
head(tree)
str(tree)
colnames(tree)
```

##Question7: Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
```{r,chunk6,echo=TRUE}
ne <- subset(tree,Site=="northeast")
head(ne)
tail(ne)
str(ne)
mean(ne$Circumf_2004_cm)
mean(ne$Circumf_2019_cm)
sd(ne$Circumf_2004_cm)
sd(ne$Circumf_2019_cm)

#repeat same procedure for the southwest 
sw <- subset(tree,Site=="southwest")
head(sw)
tail(sw)
str(sw)
mean(sw$Circumf_2004_cm)
mean(sw$Circumf_2019_cm)
sd(sw$Circumf_2004_cm)
sd(sw$Circumf_2019_cm)
```
In question 7, I first created two objects which are data frame for northeast and southwest, and then using the subset command to subset the tree growth obejcts so as to get the data that corresponds to both the northeast and southwest and then using the head and tail command to properly import and obtain the first and last 6 of the tree growths for both sites, making sure the structures are checked, and then proceed to calculate and obtain the mean and standard deviation of tree circumference at the start and end of the study for both sites using the mean and sd command

##Question8: make a boxplot of tree circumference at the start and end of the study at both sites
```{r,plot2}
boxplot(ne$Circumf_2004_cm,ne$Circumf_2019_cm,sw$Circumf_2004_cm,sw$Circumf_2019_cm,
        names =c("ne2004","ne2019","sw2004","sw2019"),ylab="Circumference (cm)" ,
        main="Tree growth at both plantation sites")
```
In question 8, i use the boxplot command to plot a chart showing the tree circumference at the start and end of the study for both sites

##Question9: Calculate the mean growth over the past 10 years at each site
```{r,chunk7,echo=TRUE}
ne$growth <-ne$Circumf_2019_cm-ne$Circumf_2009_cm
head(ne)
mean(ne$growth)

#repeat same procedure for the southwest
sw$growth <-sw$Circumf_2019_cm-sw$Circumf_2009_cm
head(sw)
mean(sw$growth)
```
Using the downloaded subset, i substract the past year which is 2009 from the recent year which is 2019 and using the head command to check if imported properly, and then using the mean command to calculate the mean growth over the past 10 years, repeated the steps for both sites.

##Question 10: use the t.test and wilcox.test functions to estimate the p-value and show that the 10 year growth is different at the two sites.
```{r,chunk8}
t.test(sw$growth,ne$growth)
wilcox.test(sw$growth,ne$growth)
```
i use the `t.test` and `wilcox.test` command to determine the difference between the two sites

# PART 2: Determine the limits of BLAST

To start with this part 2 section, i first of all, loaded all the library packages that are required in order to get all the codes working fine.
```{r,libs}
library("seqinr")
library("R.utils")
library("rBLAST")
library("ape")
library("ORFik")
library("Biostrings")
```
##Question 1: Download the whole set of E. coli gene DNA sequences and use gunzip to decompress, Use the makeblast() function to create a blast database. How many sequences are present in the E.coli set?
```{r,chunk9}
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",
              destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite= TRUE)
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa",dbtype = "nucl","-parse_seqids")
```
Using the `download.file` command to download all the E.coli gene DNA sequences, and then decompress with gunzip from the `R.utils` package, also used functions from the `rBlast` package to create a blast database and determining the number of sequences present in the E.coli set which amounts to 4140.

##Question 2: Download the sample fasta sequences and read them in as above, For your allocated sequence,determine the length (in bp) and the proportion of GC bases.
```{r,chunk10}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa",destfile = "sample.fa")
s <- read.fasta("sample.fa")
myseq <- s[[28]]
str(myseq)
seqinr::GC(myseq)
seqinr::getLength(myseq)
length(myseq)
```
Using the `download.file` command to download the sample fasta sequences and i load it into R in an object called s, and then select out my sequence from the batch of all the sample fasta sequences and checking the structures before proceeding to use the `seqinr` package to determine the length (in bp) and the proportion of GC bases.

##Question 3: You will be provided with R functions to create BLAST databases and perform blast searches, Use blast to identify what E. coli gene your sequence matches best. Show a table of the top 3 hits including percent identity, E-value and bit scores.
```{r,blast1}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R",
              destfile = "mutblast.R")
source("mutblast.R")
myblastn_tab
res <- myblastn_tab(myseq = myseq, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
str(res)
res
```
I use the `download.file` command to download the mutblast R Fuctions from Github and then i source the file so as to save the funtions in my environment and then to test the functions and also to generate an object called `res`, then i blast with the query as my allocated sequence, making sure the structures is check so as to generate a table with the `res` object initially generated. In the table we can find the top 3 hits including percent identity = 100, E-value = 0 and bit scores = 2769. The discovered gene is AAC75752 and it is an anaerobic nitric oxide reductase flavorubredoxin and the gene name is norV b2710.

##Question 4: You will be provided with a function that enables you to make a set number of point mutations to your sequence of interest. Run the function and write an R code to check the number of mismatches between the original and mutated sequence.
```{r,mutator1}
myseqmutator <-mutator(myseq, 28)
res <- myblastn_tab(myseq = myseqmutator, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
res
```
using the `mutator` to mutate the original sequence and then using the `Blastn` command to import and compare the changes in the mismatches of which from the original sequence the mismatch was 0 and when the original sequence was mutated the mismatch turns to 19 and there is a noticeable change in the bitscore and p.identity.

##Question 5: Using the provided functions for mutating and BLASTing a sequence, determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin. Because the mutation is random, you may need to run this test multiple times to get a reliable answer.
```{r,blast}
myfunc <- function(myseq,nmut) { 
  mutseq <- mutator( myseq= myseq, nmut = nmut) 
  res <- myblastn_tab(myseq= myseq,db= "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa") 
  if (is.null(res)) {myres= 0} else {myres = 1} 
  return(myres)
}
myfunc(myseq,100)
replicate(n = 100, myfunc(myseq,100))
mean(replicate(100,myfunc(myseq,100)))
n <-c(0,100,200,300,400,500,600,700,800,900,1000,1100,1200,1300,1400)
myfunction_rep <- function(nmut) {
  mean(replicate(n=100, myfunc(myseq,100)))
}
finalres <- sapply( n, myfunc_rep )
finalres
```
first I created a fucntion `myfunc` that can mutate + blast and then summarise the result as a 0 or 1 by testing the new function `myfunc`, towards being sure that the blast was successful and then I use the `replicate` command to repeat the routine a 100 times by making sure that I am getting an accurate result which gave me a vector of 100 values (1s). Towards answering the question which is "Addressing the limit of Blastn", I tested some larger numbers almost to the length of my sequence with a new function `myfunction_rep` and then concluded with the sapply command which shows a few bunch of (1s).

## 6: Provide a chart or table that shows how the increasing proportion of mutated bases reduces the ability for BLAST to match the gene of origin. Summarize the results in 1 to 2 sentences.
